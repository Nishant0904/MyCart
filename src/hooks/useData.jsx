import { useEffect, useState } from 'react'

const useData = (path) => {
    const [data, setData] = useState(null)
    const [error, setError] = useState(null)
    const [loading, setLoading] = useState(true)

    useEffect(() => {
        fetch(`https://fakestoreapi.com${path}`)
            .then((response) => {
                if (response.status >= 400) {
                    throw new Error('server error')
                }
                return response.json()
            })
            .then((response) => setData(response))
            .catch((error) => setError(error))
            .finally(() => setLoading(false))
    }, [path])

    return { data, error, loading }
}

export default useData
