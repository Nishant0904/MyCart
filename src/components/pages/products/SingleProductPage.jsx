import styled from 'styled-components'
import useData from '../../../hooks/useData'
import LoadingPage from '../loading/LoadingPage'
import ProductQuantitySelector from './ProductQuantitySelector'
import formatToUSD from '../../../utils/formatToUSD'
import PropTypes from 'prop-types'

const SingleProductPage = ({ id, handleAddToCart }) => {
    const { data, error, loading } = useData(`/products/${id}`)

    if (error) return <p>An error occurred. Please check back later.</p>
    if (loading) return <LoadingPage />

    return (
        <StyledSingleProductPage data-testid='single-product-page'>
            <SingleProductContainer>
                <ProductImageWrapper>
                    <img src={data.image} alt={data.title} />
                </ProductImageWrapper>
                <ProductInfoContainer>
                    <ProductMetaData>
                        <div>SKU: {data.id}</div>
                        <div>Category: {data.category}</div>
                    </ProductMetaData>
                    <ProductPrimaryInfo>
                        <h1 data-testid='single-product-page__product-title'>{data.title}</h1>
                        <div>{formatToUSD(data.price)}</div>
                    </ProductPrimaryInfo>
                    <p>{data.description}</p>
                    <ProductQuantitySelector
                        product={data}
                        handleAddToCart={handleAddToCart}
                    />
                </ProductInfoContainer>
            </SingleProductContainer>
        </StyledSingleProductPage>
    )
}

const StyledSingleProductPage = styled.section`
    flex-grow: 1;
    align-self: center;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 48px 96px;
    max-width: 1200px;

    @media (max-width: 740px) {
        padding: 48px 24px;
    }
`

const SingleProductContainer = styled.div`
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(336px, 1fr));
    gap: 48px;
`

const ProductImageWrapper = styled.div`
    display: flex;
    justify-content: center;

    img {
        max-width: 400px;
    }
`

const ProductInfoContainer = styled.div`
    margin-top: 48px;
    display: flex;
    flex-direction: column;
    gap: 8px;
`

const ProductMetaData = styled.div`
    display: flex;
    flex-direction: column;
    gap: 2px;
    text-transform: uppercase;
    font-weight: 900;
    color: var(--secondary-txt-color);
    font-size: 0.8rem;
    letter-spacing: 0.1rem;
`

const ProductPrimaryInfo = styled.div`
    display: flex;
    align-items: end;
    justify-content: space-between;

    div {
        font-size: 1.5rem;
    }
`

SingleProductPage.propTypes = {
    id: PropTypes.string,
    handleAddToCart: PropTypes.func.isRequired,
}

export default SingleProductPage
