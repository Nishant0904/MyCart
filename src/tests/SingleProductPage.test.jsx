import { describe, it, expect, vi } from 'vitest'
import { act } from 'react-dom/test-utils'
import { render, screen } from '@testing-library/react'
import SingleProductPage from '../components/pages/products/SingleProductPage'
import useData from '../hooks/useData'

vi.mock('../hooks/useData', () => ({
    __esModule: true,
    default: vi.fn(),
}))

const mockProduct = {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
        'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
        rate: 3.9,
        count: 120,
    },
}

describe('SingleProductPage test', () => {
    it('renders error', async () => {
        useData.mockImplementation(() => {
            return { data: null, error: true, loading: false }
        })

        await act(async () => {
            render(<SingleProductPage id={1} handleAddToCart={() => {}} />)
        })

        expect(screen.getByText('An error occurred. Please check back later.')).toBeInTheDocument()
    })

    it('renders loading page when data is null and loading is true', async () => {
        useData.mockImplementation(() => {
            return { data: null, error: null, loading: true }
        })

        await act(async () => {
            render(<SingleProductPage id={1} handleAddToCart={() => {}} />)
        })

        expect(screen.getByTestId('loading-page')).toBeInTheDocument()
    })

    it('renders product data correctly when data is available and loading is false', async () => {
        useData.mockImplementation(() => {
            return { data: mockProduct, error: null, loading: false }
        })

        await act(async () => {
            render(<SingleProductPage id={1} handleAddToCart={() => {}} />)
        })

        expect(screen.getByTestId('single-product-page')).toBeInTheDocument()
        expect(screen.getByTestId('single-product-page')).toMatchSnapshot()
    })
    
    it('renders correct product title', async () => {
        useData.mockImplementation(() => {
            return { data: mockProduct, error: null, loading: false }
        })

        await act(async () => {
            render(<SingleProductPage id={1} handleAddToCart={() => {}} />)
        })

        expect(screen.getByTestId('single-product-page__product-title').textContent).toBe(mockProduct.title)
    })
})
