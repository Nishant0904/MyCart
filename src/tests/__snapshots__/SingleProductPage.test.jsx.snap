// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SingleProductPage test > renders product data correctly when data is available and loading is false 1`] = `
<section
  class="sc-jTQCzO cNKGRn"
  data-testid="single-product-page"
>
  <div
    class="sc-gLLuof XMglp"
  >
    <div
      class="sc-iBdnpw jhlKia"
    >
      <img
        alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
        src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      />
    </div>
    <div
      class="sc-fsYfdN iHfCed"
    >
      <div
        class="sc-qZrbh hdJorP"
      >
        <div>
          SKU: 
          1
        </div>
        <div>
          Category: 
          men's clothing
        </div>
      </div>
      <div
        class="sc-jsEeTM cSiJou"
      >
        <h1
          data-testid="single-product-page__product-title"
        >
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </h1>
        <div>
          $109.95
        </div>
      </div>
      <p>
        Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday
      </p>
      <div
        class="sc-hLQSwg lcWqWs"
        data-testid="product-quantity-selector"
      >
        <div
          class="sc-eDLKkx kaRKTi"
        >
          <button
            class="sc-dmyCSP kbuipJ"
            data-testid="product-quantity__decrement-btn"
          >
            â€“
          </button>
          <input
            data-testid="product-quantity__number-input"
            type="number"
            value="1"
          />
          <button
            class="sc-dmyCSP kbuipJ"
            data-testid="product-quantity__increment-btn"
          >
            +
          </button>
        </div>
        <button
          class="sc-dmyCSP kbuipJ"
          data-testid="product-quantity__add-to-cart-btn"
        >
          Add to Cart
        </button>
      </div>
    </div>
  </div>
</section>
`;
